File: Server/Javascript/master.js

var R = require('./rl.js'); 


// helper functions
var fillMat = function(input, mat) {
	for(var i = 0; i < input.length; i++) {
		for(var j = 0; j < input[0].length; j++) {
			mat.set(i, j, input[i][j]);
		}
	}
}

// Coming soon: addons for rljs(new functions for the graph class so it can do more things)


// Main code 
function Master(socketServer, hyperparams, data, graphRep) {
	this.server = socketServer;
	this.data = data; 
	this.model = graphRep; 
	this.hyperparams = hyperparams
	this.epochs = hyperparams.epochs || 20;
	this.graphMats = {};
	this.workers = {};
	this.updateCounter = {}; 
	this.idx = 0;
	this.update = graphRep.updateWorkerCount || 1;
	this.mu = hyperparams.initMu;
	this.std = hyperparams.initStd;
	this.solver = new R.Solver();
	this.batchSize = hyperparams.batchSize || 10;
	this.lr = hyperparams.lr || 0.01;
	this.regc = hyperparams.regularization || 0.00001;
	this.clip = hyperparams.clip || 5.0; 
}


Master.prototype.createGraph = function() {
	this.graph = new R.Graph(); 
	for(var name in this.model.params){
		var node = this.model.params[name];
		this.graphMats[name] = new R.RandMat(node.nr, node.nc, this.mu, this.std);
	}
}

Master.prototype.sendWeights = function(socket) {
	var json = {};
	json.data = {};
	for(var mat in this.graphMats) {
		json.data[mat] = this.graphMats[mat].toJSON();
	}
	json.step_cache = this.solver.step_cache; 
	json.id = socket.conn.id;
	socket.emit('weights', json);
}

Master.prototype.incrementUpdateCounter = function() {
	for(var socket in this.updateCounter) {
		this.updateCounter[socket] += 1; 
	}
}

Master.prototype.sendBatch = function(socket, batch) {
	socket.emit('batch', batch);
}

Master.prototype.addSocketCallBacks = function() {
	var updateWeights = function(data) {
		for(var mat in data.grads) {
			var toUpdate = this.graphMats[mat]; 
			var update = data.grads[mat]; 
			for(var weight in toUpdate.dw) {
				toUpdate.dw[weight] = update[weight]; 
			}
		}
		this.solver.step(this.graphMats, this.lr, this.regc, this.lr);
		this.incrementUpdateCounter();
		this.model.reportCost(data.cost);
		if(this.updateCounter[data.id] >= this.update) {
			this.sendWeights(this.workers[data.id]);
			this.updateCounter[data.id] = 0; 
		}
		if(!data.stop) {
			var inputs = this.data.inputs.slice(this.idx, this.idx + this.batchSize);
			var labels = this.data.labels.slice(this.idx, this.idx + this.batchSize);
			var batch = {
				inputs: inputs,
				labels: labels
			}
			this.idx += this.batchSize
			if(this.idx + this.batchSize > this.data.inputs.length) {
				this.idx = 0
				this.epochs = this.epochs - 1;
			}
			if(this.epochs <= 0) {
				console.log(new Number(new Date()))
				console.log('done')
				return; 
			}
			this.sendBatch(this.workers[data.id], batch);
		}
	}

	var updateWeightscb = updateWeights.bind(this);

	var preProcess = function(raw) {
		console.log(raw);
		var batch = this.model.preProcess(raw.data);
		batch.stop = true;
		this.sendBatch(this.workers[raw.id], batch);
	}
	var preProcesscb = preProcess.bind(this);

	var connect = function(socket) {
		this.workers[socket.conn.id] = socket; 
		this.updateCounter[socket.conn.id] = 0; 
		this.sendWeights(socket)
		socket.on('update', updateWeightscb);
		socket.on('raw', preProcesscb);
		if(this.model.startImmediately) {
			var inputs = this.data.inputs.slice(this.idx, this.idx + this.batchSize);
			var labels = this.data.labels.slice(this.idx, this.idx + this.batchSize);
			var batch = {
				inputs: inputs,
				labels: labels
			}
			this.idx += this.batchSize
			if(this.epochs <= 0) {
				return; 
			}
			this.sendBatch(socket, batch);
		}
	}

	var disconnect = function(socket) {
		var sid = socket.conn.id; 
		if(this.workers[sid]) {
			delete this.workers[sid];
		}
		if(this.updateCounter[sid]) {
			delete this.updateCounter[sid];
		}
	}

	var connectcb = connect.bind(this);
	var disconnectcb = disconnect.bind(this);

	this.server.on('connect', connectcb);
	this.server.on('disconnect', disconnectcb);
}


Master.prototype.forward  = function(batch) {
	// vectorize inputs first
	var inputs = fillMat(batch.inputs[input], new R.Mat(this.batchSize, this.model.inputSize));
	batch.inputs = inputs
	this.model.forward(this.graph, this.graphMats, batch);
}

// Warmup Trainer
Master.prototype.train = function() {
	var batchSize = this.batchSize;
	for(var i = 0; i < this.hyperparams.warmup; i++) {
		var idx = 0; 
		while(idx < this.data.inputs.length) {
			var inputs = this.data.inputs.slice(idx, idx + this.batchSize);
			var labels = this.data.labels.slice(this.idx, this.idx + this.batchSize);
			var batch = {
				inputs: inputs,
				labels: labels
			}
			this.forward(batch);
			this.graph.backward();
			this.graph.backprop = []; 

			idx += batchSize

			if(idx + batchSize > this.data.inputs.length) {
				break;
			}
		}
	}
	console.log('done warming up')
}

module.exports = Master; 

File: Server/Python/master.py
import tensorflow as tf
import socketio
import eventlet
import json
import threading

## GraphRep represents the API any model you feed Master must satisfy
class GraphRep: 

	def __init__(self, **hyperparams):
		return
		# implement any set up you need!, probably using hyper parameters dict as kwargs

	def preprocess(self, raw):
		# for if you want to preprocess any data before giving it to workers(i.e things like word vectors)
		return

	def construct_model(self):
		return
		# implement a model using tensorflow 

	def train_batch(self, batch):
		return
		# give back the loss operation for this model(for model warmup or parallel training)

	def apply_grads(self, grads):
		return	
		# apply gradients given back by the worker to the model(this is pretty easy in tensorflow!)

	@property
	def step_cache(self):
		return
		# send back step cache(dictionary of sqrt of sum of squared gradients) 
		# so that workers can properly use RMS prop

	@property
	def params(self):
		return
		# to get all trainable vars: [v for v in tf.trainable_variables()]
		# returns a dictionary mapping each variable's name to an array of weights


class Master:
	def __init__(self, socket_server, data=None, graphRep=None, epochs=1, hyperparams=1): 
		self.socket = socket_server
		self.data = data
		self.idx = 0
		self.model = graphRep	
		self.epochs = epochs
		self.bind_socket_cbs()
		self.hyperparams = hyperparams
		self.model.construct_model()
		self.update_counter = {}

	def preprocess(self, sid, msg): 
		batch = self.graphRep.preprocess(msg)
		batch = {
			'inputs': batch['inputs'],
			'labels': batch['labels']
		}
		self.send_batch(sid, batch, stop=True)

	def send_weights(self, sid): 
		model_data = {
			'id': sid,
			'data': self.graphRep.params,
			'step_cache': self.graphRep.step_cache
		}
		self.socket.emit('weights', data=model_data, room=sid)

	# assumes data is dictionary of numpy arrays
	def send_batch(self, sid, stop=False): 
		inputs = data['inputs'][idx:idx + self.hyperparams['batch_size']].tolist()
		labels = data['labels'][idx:idx + self.hyperparams['batch_size']].tolist()
		self.idx += self.hyperparams['batch_size']
		if self.idx + self.hyperparams['batch_size'] > self.data['train'].shape[0]:
			self.idx = 0
			self.epochs -= 1

		if self.epochs <= 0:
			print 'done'
			return

		batch = {
			'inputs': inputs,
			'labels': labels,
			'stop': stop
		}
		self.socket.emit('batch', data=batch, room=sid)

	def add_socket_callbacks(self):
		self.socket.on('connect', self.connect)
		self.socket.on('disconnect', self.disconnect)
		self.socket.on('raw', self.preprocess)
		self.socket.on('weights', self.receive_updates)

	# main training loop that you should use to warm up the model
	def background_train(self):
		for epoch in xrange(self.hyperparams['epochs']): 
			while idx < len(data): 
				if idx + batch_size < len(data): 
					inputs = data['inputs'][idx:idx + self.hyperparams['batch_size']]
					labels = data['labels'][idx:idx + self.hyperparams['batch_size']]
					batch = {
						'inputs': inputs,
						'labels': labels,
					}
					self.model.train_batch(batch)
					idx += self.hyperparams['batch_size']
					self.model.apply_grads()
				else:
					break

	def increment_counts(self):
		for sid in self.update_counter:
			self.update_counter[sid] += 1

	def disconnect(self, sid, msg):
		del self.update_counter[sid]

	def connect(self, sid, msg):
		self.update_counter[sid] = 0 
		self.send_weights(sid)
		if self.hyperparams['start_immediately']:
			self.send_batch(sid)

	def receive_updates(self, sid, msg):
		self.model.apply_grads(msg)
		self.increment_counts() 
		if self.update_counter[sid] >= self.hyperparams['update_count']:
			self.send_weights(sid)
		if msg.stop:
			self.send_batch(sid, msg.stop)








